File: com/pdfjet/OptionalContentGroup.java
/*
 * OptionalContentGroup.java
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 *
 * Original author: Mark Paxton
 * Modified and adapted for use in PDFjet by Evgeni Dragoev
 */
package com.pdfjet;

import com.pdfjet.encryption.*;
import java.util.ArrayList;
import java.util.List;

/**
 * Container for drawable objects that can be drawn on a page as part of Optional Content Group.
 * Please see the PDF specification and Example_30 for more details.
 *
 * @author Mark Paxton
 */
public class OptionalContentGroup {
    /** The object number of this group. */
    protected int objNumber;
    /** The name of this group. */
    protected String name;

    private final PDF pdf;
    private int ocgNumber = -1;
    private boolean visible;
    private boolean printable;
    private boolean exportable;
    private final List<Drawable> components;

    /**
     * Creates OptionalContentGroup object
     *
     * @param pdf the PDF.
     * @param name the name of the group.
     */
    public OptionalContentGroup(PDF pdf, String name) {
        this.pdf = pdf;
        this.name = name;
        this.components = new ArrayList<Drawable>();
    }

    /**
     * Returns the name of this group.
     *
     * @return the name of the group.
     */
    public String getName() {
        return this.name;
    }

    /**
     * Add drawable object to the group
     *
     * @param drawable the drawable object
     */
    public void add(Drawable drawable) {
        components.add(drawable);
    }

    /**
     * Removes all drawable objects from the group.
     */
    public void clear() {
        components.clear();
    }

    /**
     * Returns the drawable objects in the group.
     *
     * @return the list of drawable objects.
     */
    public List<Drawable> getComponents() {
        return components;
    }

    /**
     * Sets the visibility of this group
     *
     * @param visible flag
     * @return this OptionalContentGroup object.
     */
    public OptionalContentGroup setVisible(boolean visible) {
        this.visible = visible;
        return this;
    }

    /**
     * Sets the printability of this group
     *
     * @param printable flag
     * @return this OptionalContentGroup object.
     */
    public OptionalContentGroup setPrintable(boolean printable) {
        this.printable = printable;
        return this;
    }

    /**
     * Sets the exportability of this group
     *
     * @param exportable flag
     * @return this OptionalContentGroup object.
     */
    public OptionalContentGroup setExportable(boolean exportable) {
        this.exportable = exportable;
        return this;
    }

    /**
     * Draws this content group on a page
     *
     * @param page the page to draw on
     * @throws Exception if there is a problem
     */
    public void drawOn(Page page) throws Exception {
        if (this.ocgNumber == -1) {
            pdf.newobj();
            pdf.append("<<\n");
            pdf.append("/Type /OCG\n");

            byte[] nameBytes = name.getBytes(java.nio.charset.StandardCharsets.UTF_8);
            if (pdf.encryption != null) {
                nameBytes = AES256.encrypt(nameBytes, pdf.encryption.getKey());
            }
            pdf.append("/Name <");
            pdf.append(Util.toHexString(nameBytes));
            pdf.append(">\n");

            pdf.append("/Usage <<\n");
            if (visible) {
                pdf.append("/View << /ViewState /ON >>\n");
            } else {
                pdf.append("/View << /ViewState /OFF >>\n");
            }
            if (printable) {
                pdf.append("/Print << /PrintState /ON >>\n");
            } else {
                pdf.append("/Print << /PrintState /OFF >>\n");
            }
            if (exportable) {
                pdf.append("/Export << /ExportState /ON >>\n");
            } else {
                pdf.append("/Export << /ExportState /OFF >>\n");
            }
            pdf.append(">>\n");
            pdf.append(">>\n");
            pdf.endobj();

            objNumber = pdf.getObjNumber();

            pdf.groups.add(this);
            this.ocgNumber = pdf.groups.size();
        }

        if (!components.isEmpty()) {
            page.append("/OC /OC");
            page.append(ocgNumber);
            page.append(" BDC\n");
            for (Drawable component : components) {
                component.drawOn(page);
            }
            page.append("\nEMC\n");
        }
    }
}   // End of OptionalContentGroup.java

File: src/optionalcontentgroup.go
package pdfjet

/**
 * optionalcontentgroup.go
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 *
 * Original author: Mark Paxton
 * Modified and adapted for use in PDFjet by Evgeni Dragoev
 */

import (
	"encoding/hex"
	"fmt"

	"github.com/edragoev1/pdfjet/src/encryption"
	"github.com/edragoev1/pdfjet/src/token"
)

// OptionalContentGroup is container for drawable objects that can be drawn on a page as part of Optional Content Group.
// Please see the PDF specification and Example_30 for more details.
//
// @author Mark Paxton
type OptionalContentGroup struct {
	objNumber  int
	pdf        *PDF
	name       string
	ocgNumber  int
	visible    bool
	printable  bool
	exportable bool
	components []Drawable
}

// NewOptionalContentGroup constructs optional content group.
func NewOptionalContentGroup(pdf *PDF, name string) *OptionalContentGroup {
	ocg := new(OptionalContentGroup)
	ocg.pdf = pdf
	ocg.name = name
	ocg.ocgNumber = -1
	ocg.components = make([]Drawable, 0)
	return ocg
}

// Add appends drawable component to this optional content group.
func (ocg *OptionalContentGroup) Add(drawable Drawable) {
	ocg.components = append(ocg.components, drawable)
}

// SetVisible sets the visibility of the group.
func (ocg *OptionalContentGroup) SetVisible(visible bool) {
	ocg.visible = visible
}

// SetPrintable sets the printable components.
func (ocg *OptionalContentGroup) SetPrintable(printable bool) {
	ocg.printable = printable
}

// SetExportable sets the exportable components.
func (ocg *OptionalContentGroup) SetExportable(exportable bool) {
	ocg.exportable = exportable
}

// DrawOn draws the components in the optional content group on the page.
func (ocg *OptionalContentGroup) DrawOn(page *Page) {
	if ocg.ocgNumber == -1 {
		ocg.pdf.newobj()
		ocg.pdf.appendByteArray(token.BeginDictionary)
		ocg.pdf.appendString("/Type /OCG\n")

		nameBytes := []byte(ocg.name)
		if ocg.pdf.encryption != nil {
			var err error
			nameBytes, err = encryption.Encrypt(nameBytes, ocg.pdf.encryption.GetKey())
			if err != nil {
				fmt.Println("encryption failed:", err)
				return
			}
		}
		ocg.pdf.appendString("/Name <")
		ocg.pdf.appendString(hex.EncodeToString(nameBytes))
		ocg.pdf.appendString(">\n")

		ocg.pdf.appendString("/Usage <<\n")
		if ocg.visible {
			ocg.pdf.appendString("/View << /ViewState /ON >>\n")
		} else {
			ocg.pdf.appendString("/View << /ViewState /OFF >>\n")
		}
		if ocg.printable {
			ocg.pdf.appendString("/Print << /PrintState /ON >>\n")
		} else {
			ocg.pdf.appendString("/Print << /PrintState /OFF >>\n")
		}
		if ocg.exportable {
			ocg.pdf.appendString("/Export << /ExportState /ON >>\n")
		} else {
			ocg.pdf.appendString("/Export << /ExportState /OFF >>\n")
		}
		ocg.pdf.appendString(">>\n")
		ocg.pdf.appendByteArray(token.EndDictionary)
		ocg.pdf.endobj()

		ocg.objNumber = ocg.pdf.getObjNumber()

		ocg.pdf.groups = append(ocg.pdf.groups, ocg)
		ocg.ocgNumber = len(ocg.pdf.groups)
	}

	if len(ocg.components) > 0 {
		page.appendString("/OC /OC")
		page.appendInteger(ocg.ocgNumber)
		page.appendString(" BDC\n")
		for _, component := range ocg.components {
			component.DrawOn(page)
		}
		page.appendString("\nEMC\n")
	}
}
